static/js/message_events.js

```diff
@@ -192,12 +215,22 @@ exports.update_messages = function update_messages(events) {
                     topic_name: msg.topic,
                 });
 
-                // Update the unread counts; again, this must be called
-                // before we call set_message_topic.
-                unread.update_unread_topics(msg, event);
+                if (topic_edited) {
+                    // Update the unread counts; again, this must be called
+                    // before we call set_message_topic.
+                    unread.update_unread_topics(msg, event);
+
+                    msg.topic = new_topic;
+                    msg.topic_links = event.topic_links;
+                }
+
+                if (stream_changed) {
+                    msg.stream_id = event.new_stream_id;
 
-                msg.topic = new_topic;
-                msg.topic_links = event.topic_links;
+                    // Update the unread counts; again, this must be called
+                    // before we call set_message_topic.
+                    unread.update_unread_moved_topic(msg);
+                }
```

Comment:
These topic/stream blocks can probably be combined to just having the `msg.stream_id` line be not common code, after applying the other suggestions I've made.